staging/src/k8s.io/apiserver/pkg/registry/rest/create.go

```diff
@@ -150,16 +156,24 @@ func CheckGeneratedNameError(strategy RESTCreateStrategy, err error, obj runtime
 		return err
 	}
 
-	objectMeta, kind, kerr := objectMetaAndKind(strategy, obj)
+	objectMeta, gvk, kerr := objectMetaAndKind(strategy, obj)
 	if kerr != nil {
 		return kerr
 	}
 
 	if len(objectMeta.GetGenerateName()) == 0 {
+		// If we don't have a generated name, return the original error (AlreadyExists).
+		// When we're here, the user picked a name that is causing a conflict.
 		return err
 	}
 
-	return errors.NewServerTimeoutForKind(kind.GroupKind(), "POST", 0)
+	// If we have a name and generated name, the server picked a name
+	// that already exists.
+	return errors.NewConflict(
```

Comment:
I think the old line was a copy-paste error. I would expect we'd return AlreadyExists if the generated (or provided) name matched an existing object.